libs/vkd3d/command.c

```diff
@@ -6287,7 +6355,77 @@ static void STDMETHODCALLTYPE d3d12_command_list_ClearUnorderedAccessViewFloat(d
 static void STDMETHODCALLTYPE d3d12_command_list_DiscardResource(d3d12_command_list_iface *iface,
         ID3D12Resource *resource, const D3D12_DISCARD_REGION *region)
 {
-    FIXME_ONCE("iface %p, resource %p, region %p stub!\n", iface, resource, region);
+    struct d3d12_command_list *list = impl_from_ID3D12GraphicsCommandList(iface);
+    struct d3d12_resource *texture = unsafe_impl_from_ID3D12Resource(resource);
+    VkImageSubresourceLayers vk_subresource_layers;
+    VkImageSubresource vk_subresource;
+    D3D12_RECT full_rect;
+    int attachment_idx;
+    bool full_discard;
+    unsigned int i;
+
+    TRACE("iface %p, resource %p, region %p.\n", iface, resource, region);
+
+    /* This method is only supported on DIRECT and COMPUTE queues,
+     * but we only implement it for render targets, so ignore it
+     * on compute. */
+    if (list->type != D3D12_COMMAND_LIST_TYPE_DIRECT)
+    {
+        WARN("Not supported for queue type %d.\n", list->type);
+        return;
+    }
+
+    /* Ignore buffers */
+    if (!d3d12_resource_is_texture(texture))
+        return;
+
+    /* D3D12 requires that the texture is either in render target
+     * state or in depth-stencil state depending on usage flags */
+    if (!(texture->desc.Flags & (D3D12_RESOURCE_FLAG_ALLOW_RENDER_TARGET | D3D12_RESOURCE_FLAG_ALLOW_DEPTH_STENCIL)))
+    {
+        WARN("Not supported for resource %p.\n", resource);
+        return;
+    }
+
+    /* We can't meaningfully discard sub-regions of an image. If rects
+     * are specified, all specified subresources must have the same
+     * dimensions, so just base this off the first subresource */
+    if (!(full_discard = !region->NumRects))
+    {
+        vk_subresource = d3d12_resource_get_vk_subresource(texture, region->FirstSubresource, false);
+        full_rect = d3d12_get_image_rect(texture, vk_subresource.mipLevel);
+
+        for (i = 0; i < region->NumRects && !full_discard; i++)
```

Comment:
This seems a bit fishy, I would expect to see full_discard be initialized to true here, and if there is any non-full discard, it nopes out.

Reply:
The way I read the code now, it's considered a full discard if just one rect is full.

Reply:
> it's considered a full discard if just one rect is full.

Yes, that's the idea and also done this way for clears. Discarding sub-rects doesn't have any effect when the entire subresource is cleared.

Reply:
Oh, wait. All subrects apply to all subresources?

Reply:
Yes, that's why the API requires to have all subresources to have the same dimension if discard rects are used.